### PR TITLE
Add metric query mechanism

### PR DESCRIPTION
**Pull request by jtaleric**
- Query Prometheus if it is provided to collect system information

### CODE REVIEW

File: metrics/system.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+func PromCheck(url string) bool {
+	_, ret := promQueryRange(time.Now(), time.Now(), "prometheus_build_info", url)
+	return ret
+}
+
+func QueryNodeCPU(node string, url string, start time.Time, end time.Time) (NodeCPU, bool) {
+	cpu := NodeCPU{}
+	query := fmt.Sprintf("(avg by(mode) (rate(node_cpu_seconds_total{instance=~\"%s:.*\"}[1m])) * 100)", node)
```
Comment (rsevilla87) [inline metrics/system.go:43]: s/1m/2m

File: metrics/system.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+			if strings.Contains(s.Metric.String(), "system") {
+				cpu.System = avg(s.Values)
+			}
+			if strings.Contains(s.Metric.String(), "user") {
+				cpu.User = avg(s.Values)
+			}
+			if strings.Contains(s.Metric.String(), "nice") {
+				cpu.Nice = avg(s.Values)
+			}
+			if strings.Contains(s.Metric.String(), "\"irq\"") {
```
Comment (rsevilla87) [inline metrics/system.go:67]: Is this correct? `"\"irq\""`

File: metrics/system.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	val, q := promQueryRange(start, end, query, url)
+	if !q {
+		logging.Error("Issue querying Prometheus")
+		return pods, false
+	}
+	if val.Type() == model.ValMatrix {
+		v := val.(model.Matrix)
+		for _, s := range v {
+			p := PodCPU{
+				Name:  re.FindStringSubmatch(s.Metric.String())[1],
```
Comment (rsevilla87) [inline metrics/system.go:95]: I think you can use get the pod name by using the notation `s.Metric["pod"]`

File: metrics/system.go (hunk 2 above)
Comment (jtaleric) [reply metrics/system.go:67]: Yes, because if I have `irq` it will pick up both `softirq` and `irq` having the `"` will isolate to just irq.

**Comment by jtaleric** [comment]
Just a note @rsevilla87 I haven't tested this against OCP, just Kind+Prom...

File: metrics/system.go (hunk 3 above)
Comment (jtaleric) [reply metrics/system.go:95]: :nod

**Comment by dry923** [comment]
@rook cant we move the npr decleration in netperf.go from line 122-125 and 178-181 up to the start of the for loop in line 107? Seems like it doesnt need to be in there twice since its only going to do one of those to given the if/else statement

File: metrics/system.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		}
+		token, err := client.CoreV1().ServiceAccounts("openshift-monitoring").CreateToken(context.TODO(), "prometheus-k8s", &request, v1.CreateOptions{})
+		if err != nil {
+			return conn, false
+		}
+		conn.Token = token.Status.Token
+		logging.Debugf("Token : %s", conn.Token)
+		conn.OpenShift = true
+	}
+	if !found {
```
Comment (dry923) [inline metrics/system.go:110]: Why are we doing 2 if !found statements in a row? Also it feels like we handle the !found case on line 78 unless i'm misreading it. Couldnt we just return there?

Comment (jtaleric) [reply metrics/system.go:111]: yeah, I can reduce the conditionals.

**Comment by mohit-sheth** [comment]
having a metrics profile be more customizable ...for our regular testing we can always use the above queries though

**Comment by jtaleric** [comment]
> > having a metrics profile be more customizable ...for our regular testing we can always use the above queries though
> 
> agreed, but since this is a specific tool, serving a specific need, having a more opinionated solution I think is more appropriate.
> 
> I do like having a customizable option later though...

Let me clarify a bit. when I mean customize option, I mean, we have specific queries that the user can customize -- ie they don't care about node CPU so maybe they just get the top5 pod CPU utilization.

**Review by mohit-sheth — APPROVED** [review]
with my limited go knowledge, lgtm

**Comment by jtaleric** [comment]
> @rook cant we move the npr decleration in netperf.go from line 122-125 and 178-181 up to the start of the for loop in line 107? Seems like it doesnt need to be in there twice since its only going to do one of those to given the if/else statement

Russ and I discussed this - this can be followed up on in a different patchset.

File: netperf.go [hunk 5]
```diff
@@ -113,13 +145,21 @@ func main() {
 					npr.ThroughputSummary = append(npr.ThroughputSummary, nr.Throughput)
 					npr.LatencySummary = append(npr.LatencySummary, nr.Latency99ptile)
 				}
+				npr.EndTime = time.Now()
+				if pavail {
+					npr.ClientMetrics, _ = metrics.QueryNodeCPU(s.ClientNodeInfo, pcon, npr.StartTime, npr.EndTime)
```
Comment (rsevilla87) [inline netperf.go:150]: This code is repeated 3 times, I think it's better to move it to a function

Comment (jtaleric) [reply netperf.go:150]: i am sure I can reduce it 👍🏽

**Pull request merged**